plugin/storage/es/spanstore/writer_test.go

```diff
@@ -292,21 +293,69 @@ func TestWriteSpanInternalError(t *testing.T) {
 		indexName := "jaeger-1995-04-21"
 		indexService.On("Index", stringMatcher(indexName)).Return(indexService)
 		indexService.On("Type", stringMatcher(spanType)).Return(indexService)
-		indexService.On("BodyJson", mock.AnythingOfType("*spanstore.Span")).Return(indexService)
+		indexService.On("BodyJson", mock.AnythingOfType("**dbmodel.Span")).Return(indexService)
 		indexService.On("Add")
 
 		w.client.On("Index").Return(indexService)
 
-		jsonSpan := &json.Span{
-			TraceID: json.TraceID("1"),
-			SpanID:  json.SpanID("0"),
+		jsonSpan := &dbmodel.Span{
+			TraceID: dbmodel.TraceID("1"),
+			SpanID:  dbmodel.SpanID("0"),
 		}
 
 		w.writer.writeSpan(indexName, jsonSpan)
 		indexService.AssertNumberOfCalls(t, "Add", 1)
 	})
 }
 
+func TestNewSpanTags(t *testing.T) {
+	client := &mocks.Client{}
+	logger, _ := testutils.NewLogger()
+	metricsFactory := metrics.NewLocalFactory(0)
+	testCases := []struct {
+		writer   *SpanWriter
+		expected dbmodel.Span
+	}{
+		{
+			writer: NewSpanWriter(SpanWriterParams{Client: client, Logger: logger, MetricsFactory: metricsFactory,
+				AllTagsAsFields: true}),
+			expected: dbmodel.Span{Tag: map[string]interface{}{"foo": "bar"}, Tags: []dbmodel.KeyValue{},
+				Process: &dbmodel.Process{Tag: map[string]interface{}{"bar": "baz"}, Tags: []dbmodel.KeyValue{}}},
+		},
+		{
+			writer: NewSpanWriter(SpanWriterParams{Client: client, Logger: logger, MetricsFactory: metricsFactory,
+				TagKeysAsFields: []string{"foo", "bar", "rere"}}),
+			expected: dbmodel.Span{Tag: map[string]interface{}{"foo": "bar"}, Tags: []dbmodel.KeyValue{},
+				Process: &dbmodel.Process{Tag: map[string]interface{}{"bar": "baz"}, Tags: []dbmodel.KeyValue{}}},
+		},
+		{
+			writer: NewSpanWriter(SpanWriterParams{Client: client, Logger: logger, MetricsFactory: metricsFactory}),
+			expected: dbmodel.Span{
+				Tags: []dbmodel.KeyValue{{
+					Key:   "foo",
+					Type:  dbmodel.StringType,
+					Value: "bar",
+				}},
+				Process: &dbmodel.Process{Tags: []dbmodel.KeyValue{{
+					Key:   "bar",
+					Type:  dbmodel.StringType,
+					Value: "baz",
+				}}}},
+		},
+	}
+
+	s := &model.Span{Tags: []model.KeyValue{{Key: "foo", VStr: "bar"}},
+		Process: &model.Process{Tags: []model.KeyValue{{Key: "bar", VStr: "baz"}}}}
+	for _, test := range testCases {
+		mSpan := test.writer.spanConverter.FromDomainEmbedProcess(s)
```

Comment:
Wrap every iteration with its own t.Run() with an appropriate name. In case of a failure, it should be easy to identify what's the combination that failed.